File: 프로그래머스/2/42587. 프로세스/프로세스.py
from collections import deque

def solution(priorities, location):
    answer = 0
    
    n = len(priorities)
    
    target_process = deque([False] * n)
    target_process[location] = True
    
    priorities = deque(priorities)
    
    run_process = max(priorities)
    
    while priorities:
        process, is_target = priorities.popleft(), target_process.popleft()
        
        if process >= run_process:
            answer += 1
            
            if is_target:
                return answer
            
            run_process = max(priorities)
        else:      
            priorities.append(process)
            target_process.append(is_target)
    

File: 프로그래머스/2/42587. 프로세스/프로세스.kt
import java.util.ArrayDeque

class Solution {
    fun solution(priorities: IntArray, location: Int): Int {
        var answer = 1
        val queue = ArrayDeque<Pair<Int, Int>>()
        
        for ((i, p) in priorities.withIndex()) {
            queue.add(Pair(i, p))
        }
        
        while (queue.isNotEmpty()) {
            val (l, p) = queue.removeFirst()
            val maxP = queue.maxOfOrNull {it.second} ?: 0
            
            if (p < maxP) {
                queue.add(Pair(l, p))
            } else {
                if (l == location) {
                    return answer
                }
                
                answer++
            }
        }
        
        error("종료")
    }
}